File: src/Striver_SDE_Sheet/Stack_and_Queue/NextGreaterElement/Greater_Element.cpp
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        vector<int> res(nums1.size(), -1);
        stack<int> st;
        unordered_map<int, int> umap;
        for(int i=0; i<nums2.size(); i++) {
            int element = nums2[i];
            while(!st.empty() && element > st.top())
            {	
                umap[st.top()] = element;
                st.pop();
            }   
            st.push(element);
        }
        for(int i=0; i<nums1.size(); i++) {
            int ele = nums1[i];
            if(umap.find(ele) != umap.end())
            {
                int nge = umap[ele];
                res[i] = nge; 
            }       
        }
        return res;
    }
};

File: src/Striver_SDE_Sheet/Stack_and_Queue/NextGreaterElement/Greater_Element.java
package Striver_SDE_Sheet.Stack_and_Queue.GreaterElement;
import java.util.*;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        int[] result = new int[nums1.length];
        Stack<Integer> stack=new Stack<>();
        Map<Integer, Integer> map = new HashMap<>();

        for(int num: nums2){            
            while(!stack.isEmpty() && num > stack.peek())
                map.put(stack.pop(), num);
            stack.push(num);            
        }
        
        int i=0;        
        for(int num : nums1)
            result[i++] = map.getOrDefault(num, -1);
        return result;
    }
}